benchmark/single-source/Flatten.swift.gyb

```diff
@@ -0,0 +1,275 @@
+//===--- Flatten.swift -------------------------------------  -*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+% # Ignore the following warning. This _is_ the correct file to edit.
```

Comment:
Please don't add gyb files. Eventually we want to move the swift bm suite to swiftpm and then we cannot support gyb files anymore. Beside that, gyb files make it harder to understand and maintain the code.

Reply:
I’m really surprised by this perspective. I find that to get proper performance test coverage which tests majority of legal combinations from stdlib types and methods, GYB is indispensable especially for the for ease of maintenance. The only alternative is copy & paste of some template and manual modification, which would be much more error prone.

Generally speaking, currently Swift’s biggest performance weakness is optimizer fragility. Small variations in expressing functionally  equivalent code often leed to unexpected pitfalls with orders of magnitude worse performance. My idea to harden the optimizer is to produce broad benchmark coverage and file bugs for all the gotchas. Is that a bad approach?

I’m currently working on reintroducing Existential benchmark family and the first step was to create a `.gyb` that let’s me do same refactoring without tearing my hair out in the 800 LOC file.

Reply:
There are 2 things here:
1) gyb: I'm strongly against adding more gyb files for the reasons I explained above. Eventually we will also eliminate the existing gyb files (and BTW, that's also what the stdlib team is doing).

2) Adding a lot of benchmarks just to cover all combinations is a problematic approach and not workable in the long term, because it will result in long benchmark runtimes. In general I would prefer that we add fewer but more "relevant" benchmarks (whatever this means). For example, simple operations, which we expect to be optimized to a trivial code pattern, are ofter better tested with a lit test than a benchmark. On the other hand, complex operations often don't need many combinations to be benchmarked, e.g. stdlib's sort can be benchmarked with a few types/array sizes - there is not much value in exploding the whole problem space.

But: what can be done is to add a large set of benchmarks for a specific feature, which are disabled by default (with tags). Whenever someone works on that feature he/she can run this large set locally to fully verify the performance.

Reply:
> Eventually we want to move the swift bm suite to swiftpm and then we cannot support gyb files anymore.

Building benchmarks with swiftpm does not technically conflict with use of GYB, because we are not generating the swift files as a part of cmake build. These are generated manually by running `generate_harness.py`. That  is why we are also committing the generated `.swift` files to git, design mandated by @gottesmm when we added the support for GYB, specifically to accommodate the future swiftpm builds.

> Eventually we will also eliminate the existing gyb files (and BTW, that's also what the stdlib team is doing).

The stdlib team is doing that, because the improved expressivity of Swift has obviated the need to generate a lot of boilerplate. This argument does not apply to maintenance of benchmark families build from common template that need to be parametrized for different variants.

> Adding a lot of benchmarks just to cover all combinations is a problematic approach and not workable in the long term, because it will result in long benchmark runtimes.

This is simply not true. See #20666:
> New BenchmarkCategory.existential was added to tag these tests. Running these 108 test in a manner equivalent to run_smoke_bench (time ./Benchmark_O --tags=existential --sample-time=0.0025 --num-samples=3) take 1.3 seconds on my 2008 MBP — properly sized benchmarks improve measurement precision and have negligible impact on the overall time it takes to run benchmark suite.

Reply:
@eeckstein:
> Adding a lot of benchmarks just to cover all combinations is a problematic approach and not workable in the long term, because it will result in long benchmark runtimes.

This whole new `Flatten` family at the current state (before trimming all the fat):
````
$ time ./Benchmark_O --num-samples=3 --sample-time=0.0025 {319..386}
…
Total performance tests executed: 68

real	0m1.298s
user	0m0.949s
sys	0m0.045s
````
On a 2008 MBP.